src/compile/kinds/compiler.ts

```diff
@@ -128,16 +119,35 @@ export async function getCompilerForVersion<T extends CompilerMapping>(
     );
 
     if (!fse.existsSync(compilerLocalPath)) {
-        const response = await axios({
-            method: "GET",
-            url: `${BINARIES_URL}/${prefix}/${compilerFileName}`,
-            responseType: "stream"
+        const build = md.builds.find((b) => b.version === version);
+
+        assert(
+            build !== undefined,
+            `Unable to find build metadata for ${prefix} compiler ${version} in "list.json"`
+        );
+
+        const response = await axios.get<Buffer>(`${BINARIES_URL}/${prefix}/${compilerFileName}`, {
+            responseType: "arraybuffer"
         });
 
-        const target = fse.createWriteStream(compilerLocalPath, { mode: 0o555 });
-        const pipeline = promisify(stream.pipeline);
+        const hash = crypto.createHash("sha256");
+
+        hash.update(response.data);
+
+        const digest = "0x" + hash.digest("hex");
+
+        assert(
+            digest === build.sha256,
+            `Downloaded ${prefix} compiler ${version} hash ${digest} does not match hash ${build.sha256} from "list.json"`
+        );
+
+        /**
+         * Native compilers are exeсutable files, so give them valid permissions.
+         * WASM compilers are loaded by NodeJS, so write them as common files.
+         */
+        const options = kind === CompilerKind.Native ? { mode: 0o555 } : undefined;
```

Comment:
According to the writeFile [docs](https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback) the default permissions for a new file are r/w - 0o666. Can we make it read-only (0o444) just in case?

Reply:
For wasm modules that is

Reply:
Addressed in https://github.com/ConsenSys/solc-typed-ast/pull/97/commits/bc60c2c4f19f074ef26fc2875d4e4b404b218c9f